Cleanup `MegaBlocks` Installation

eitanturok opened the pull request:
**Update 2024-07-25:**

1. We will use `stk` for now and not my fork. We'll take care of this in a separate PR.
2. We will update the repo author and email in a separate PR.
3. I added a `_version.py` for versioning and we dynamically read the version in `setup.py`.


-----

This PR cleans up the `MegaBlocks` installation process. Specifically, it
1. Adds a `pyproject.toml` file
4. Updates the `README.md` installation instructions
5. Ignores more files in `.gitignore`
6. Adds missing packages to `setup.py`, i.e. `torch` and `numpy`
7. Adds missing packages for testing to `megablocks[dev]`, i.e. `coverage`, `pytest`, `pytest_codeblocks`, `pytest-cov`, `pre-commit`. 
8. Gracefully handles missing `torch` while running `setup.py`
9. Fix style in `setup.py`: replade `'` with `"` and add license header
10. Improve `long_description` by parsing `README.md` better (taken from `composer`)
11. `MegaBlocks` depends on `stk`. `stk`'s `setup.py` unnecessarily imports torch which causes issues with installing `stk` before we have `torch` installed. We now install my [fork](https://github.com/eitanturok/stk) of `stk` that fixes this issue while waiting for @tgale96 to merge my [PR](https://github.com/stanford-futuredata/stk/pull/13) fixing this in `stk`.

b-chu commented on setup.py:
```diff
@@ -1,67 +1,139 @@
+else:
+    warnings.warn('Warning: No CUDA devices; cuda code will not be compiled.')
+
+###############################################################################
```
Remove extra README

eitanturok replied:
Removed.

b-chu commented on setup.py:
```diff
@@ -1,67 +1,139 @@
+extra_deps['all'] = list(
+    set(dep for deps in extra_deps.values() for dep in deps))
+
+###############################################################################
+# Extension Modules
+###############################################################################
```
Follow Composer style

eitanturok replied:
Removed these comments.

b-chu commented on setup.py:
```diff
@@ -1,67 +1,139 @@
+    'numpy>=1.21.5,<2.1.0',
+    'torch>=2.3.0,<2.4',
+    'triton>=2.1.0',
+    # 'stanford-stk==0.7.0',
```
Why are we versioning on your fork?

b-chu commented on pyproject.toml:
Base off composer and add C++ changes as necesary

eitanturok replied:
Addressed.

b-chu commented on setup.py:
```diff
@@ -1,67 +1,139 @@
-]
+# We require torch in setup.py to build cpp extensions "ahead of time"
+# More info here: # https://pytorch.org/tutorials/advanced/cpp_extension.html
+is_torch_installed = False
```
Unused

eitanturok replied:
Removed.

b-chu commented on .gitignore:
Where is this taken from?

eitanturok replied:
Took Composer's .gitignore

b-chu reviewed (commented):
Thanks for the PR!

eitanturok commented on pyproject.toml:
```diff
@@ -0,0 +1,504 @@
+[tool.isort]
+multi_line_output = 0
+line_length = 80
+skip = [ "env", "wandb", "runs", "build", "node_modules", "examples/inference-deployments/mpt/mpt_7b_ft_handler.py" ]
```
Removed irrelevant code snippets.

eitanturok commented:
I addressed all of your comments @b-chu except for the one about using my fork of `stk`.

My fork enables us to set up github actions for MegaBlocks, something we do not need to worry about right now. So I'm going to just use the regular `stk` version so we can merge this.

Thoughts?

eitanturok commented on megablocks/__init__.py:
```diff
@@ -1,2 +1,12 @@
+from megablocks._version import __version__
+from megablocks.layers import dmoe, moe
+
+"""Some key classes are available directly in the ``MegaBlocks`` namespace."""
```
This docstring is taken from composer. I think we should keep things consistent with the composer style.

eitanturok commented on setup.py:
```diff
@@ -1,67 +1,151 @@
-    long_description=open('README.md').read(),
+    name=_PACKAGE_NAME,
+    version=repo_version,
+    author='Trevor Gale',
+    author_email='tgale@stanford.edu',
```
I'm waiting for Kobe to get back me about updating this.

eitanturok commented:
Are we ready to merge? Any other reviewers we should notify?

dakinggg commented on README.md:
```diff
@@ -22,6 +22,8 @@ NOTE: This assumes you have `numpy` and `torch` installed.
 
 Installing `megablocks[gg]` enables dMoE computation with grouped GEMM. This feature is enabled by setting the `mlp_impl` argument to `grouped`. This is currently our recommended path for Hopper-generation GPUs.
 
+To contribute to MegaBlock, install `megablocks[dev]`. Run `pre-commit install` to configure the [pre-commit](https://pre-commit.com/) hook and then run `pre-commit run` before each commit to automatically format the code.
```
```suggestion
To contribute to MegaBlocks, install `megablocks[dev]`. Run `pre-commit install` to configure the [pre-commit](https://pre-commit.com/) hook and then run `pre-commit run` before each commit to automatically format the code.
```

eitanturok replied:
Addressed.

karan6181 commented:
Tagging @mvpatel2000 for the review before we merge this PR.

dakinggg commented on setup.py:
```diff
@@ -1,67 +1,151 @@
+                                           CUDAExtension,)
+except ModuleNotFoundError as e:
+    raise ModuleNotFoundError(
+        "No module named 'torch'. `torch` is required to install this repo."
```
```suggestion
        "No module named 'torch'. `torch` is required to install megablocks."
```

eitanturok replied:
Resolved.

dakinggg reviewed (approved):
LGTM, but wanted to note that since this PR includes a `pyproject.toml`, I think it enables build isolation which may make the package more difficult to install. Since the torch dep is specified in pyproject.toml it should be ok, but upgrading to latest torch versions will need to be done with care.

karan6181 commented on megablocks/__init__.py:
```diff
@@ -1,2 +1,12 @@
-import megablocks.layers.dmoe
-import megablocks.layers.moe
+# Copyright 2024 MosaicML MegaBlocks authors
```
Do you have a plan to include the [header file](https://github.com/mosaicml/llm-foundry/blob/main/.pre-commit/FILE_HEADER) either in this PR or next? This will automatically add the copyright header when running the pre-commit.

eitanturok replied:
I'm doing this in a follow up PR.

eitanturok commented:
Currently using custom forks of packages to resolve installation issues:
* https://git@github.com/eitanturok/stk.git for stanford-stk
* git+https://git@github.com/eitanturok/grouped_gemm.git for grouped_gemm

Blocked until these are merged. @dakinggg

eitanturok commented:
Closing this PR since #127 changed a lot of things.
